geomstats/geometry/stratified/graph_space.py

```diff
@@ -381,141 +571,474 @@ def dist(self, graph_a, graph_b):
+
+    Attributes
+    ----------
+    perm_ : array-like, shape=[...,n]
+        Node permutations where in position i we have the value j meaning
+        the node i should be permuted with node j.
+    """
+
+    def __init__(self):
+        self.perm_ = None
+
+    @abstractmethod
+    def align(self, geodesic, x):
+        raise NotImplementedError("Not implemented")
+
+    @abstractmethod
+    def dist(self, geodesic, x):
+        raise NotImplementedError("Not implemented")
+
+    def _get_n_points(self, x):
+        return 1 if gs.ndim(x) == 2 else gs.shape(x)[0]
+
+    def _permute(self, metric, graph_to_permute, perm):
+        return metric.space.permute(graph_to_permute, perm)
+
+
+class PointToGeodesicAligner(_BasePointToGeodesicAligner):
+    r"""Class for the Alignment of the points with respect to a geodesic.
+
+    Implementing the algorithm in [Huckemann2010] to select an optimal alignment to a
+    point with respect to a geodesic. The algorithm sample discrete set of n_points
+    along the geodesic between [s_min, s_max] and find the permutation that gets closer
+    to the datapoints along the geodesic.
+
+    Parameters
+    ----------
+    metric : GraphSpaceMetric
+    s_min : float.
+    s_max: float.
+    n_points : int.
+
+    References
+    ----------
+    .. [Calissano2020]  Calissano, A., Feragen, A., Vantini, S.
+        “Graph Space: Geodesic Principal Components for a Population of
+        Network-valued Data.” Mox report 14, 2020.
+        https://mox.polimi.it/reports-and-theses/publication-results/?id=855.
+    .. [Huckemann2010] Huckemann, S., Hotz, T., Munk, A.
+        "Intrinsic shape analysis: Geodesic PCA for Riemannian manifolds modulo
+        isometric Lie group actions." Statistica Sinica, 1-58, 2010.
+    """
+
+    def __init__(self, s_min, s_max, n_points=10):
+        super().__init__()
+        self.s_min = s_min
+        self.s_max = s_max
+        self.n_points = n_points
+        self._s = None
+
+    def __setattr__(self, attr_name, value):
+        r"""Set attributes."""
+        if attr_name in ["s_min", "s_max", "n_points"]:
+            self._s = None
+
+        return object.__setattr__(self, attr_name, value)
+
+    def _discretize_s(self):
+        r"""Compute the domain distretization."""
+        return gs.linspace(self.s_min, self.s_max, num=self.n_points)
+
+    @property
+    def s(self):
+        r"""Save the domain distretization."""
+        if self._s is None:
+            self._s = self._discretize_s()
+
+        return self._s
+
+    def _get_gamma_s(self, geodesic):
+        r"""Evaluate the geodesic in s."""
+        return geodesic(self.s)
+
+    def _compute_dists(self, metric, geodesic, x):
+        gamma_s = self._get_gamma_s(geodesic)
+
+        n_points = self._get_n_points(x)
+        if n_points > 1:
+            gamma_s = gs.repeat(gamma_s, n_points, axis=0)
+            rep_x = gs.concatenate([x for _ in range(self.n_points)])
```

Comment:
why not using `repeat`  or `tile`?

Reply:
We basically want `gamma_s` to be repeated with repeat, which means e.g. `gamma_s = [1., 2.]` becomes `[1., 1., 2., 2.]`, and concatenate of x because e.g `x = [1., 2.]` becomes `[1., 2., 1., 2.]`. This way we have all against all.